pkg/ruler/remotequerier.go

```diff
@@ -188,14 +189,38 @@ func (q *RemoteQuerier) Query(ctx context.Context, qs string, t time.Time) (prom
 	logger, ctx := spanlogger.NewWithLogger(ctx, q.logger, "ruler.RemoteQuerier.Query")
 	defer logger.Span.Finish()
 
-	valTyp, res, err := q.query(ctx, qs, t, logger)
+	return q.query(ctx, qs, t, logger)
+}
+
+func (q *RemoteQuerier) query(ctx context.Context, query string, ts time.Time, logger log.Logger) (promql.Vector, error) {
+	req, err := q.createRequest(ctx, query, ts)
 	if err != nil {
-		return nil, err
+		return promql.Vector{}, nil
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, q.timeout)
+	defer cancel()
+
+	resp, err := q.sendRequest(ctx, &req)
+	if err != nil {
+		level.Warn(logger).Log("msg", "failed to remotely evaluate query expression", "err", err, "qs", query, "tm", ts)
+		return promql.Vector{}, err
+	}
+	if resp.Code/100 != 2 {
+		return promql.Vector{}, httpgrpc.Errorf(int(resp.Code), "unexpected response status code %d: %s", resp.Code, string(resp.Body))
 	}
-	return decodeQueryResponse(valTyp, res)
+	level.Debug(logger).Log("msg", "query expression successfully evaluated", "qs", query, "tm", ts)
+
+	contentTypeHeader := getHeader(resp, "Content-Type")
```

Comment:
Note to reviewers: the query-frontend has always set `Content-Type` to `application/json`, so checking it like this with no fallback does not introduce any backwards-compatibility issues.